src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -612,6 +612,39 @@ int D4NFilterObject::set_head_obj_dir_entry(const DoutPrefixProvider* dpp, optio
     } else {
       ldpp_dout(dpp, 10) << "D4NFilterObject::" << __func__ << "(): BlockDirectory get method failed for head object with ret: " << ret << dendl;
     }
+  } else if (!this->get_bucket()->versioned()) {
+    rgw::d4n::CacheObj null_object = rgw::d4n::CacheObj{
+      .objName = "_:null_" + this->get_name(),
+      .bucketName = this->get_bucket()->get_name(),
+      .dirty = dirty,
+      .hostsList = { dpp->get_cct()->_conf->rgw_d4n_l1_datacache_address },
+    };
+
+    rgw::d4n::CacheBlock null_block = rgw::d4n::CacheBlock{
+      .cacheObj = null_object,
+      .blockID = 0,
+      .version = this->get_object_version(),
+      .size = 0,
+    };
+
+    ret = blockDir->get(dpp, &null_block, y);
```

Comment:
Is there really a need for a 'get' and then 'set', can we just not overwrite the old values, because the most recent one is valid, we do not need to update the hostsList etc also, as this block found in the cache belonging to the local node is the only valid "null" block, so I believe its safe to just overwrite the existing values.

Reply:
I agree, but can this not occur for the versioned head object logic above as well? Either the block is set from scratch or only certain fields are updated with new values, which will happen regardless.

Reply:
Well for the versioned head object you had made these changes while storing the prevVersion and deleteMarker fields. Now that I look at this line of code: version_block.prevVersion = {version_block.version, version_block.deleteMarker}; it appears incorrect to me, shouldn't prevVersion be set to block.version before setting the new version for block? I am not sure whether you had any other reason in mind for making a get and set call for a versioned head block, but it appears that you only need to do a set and also correct the prevVersion field value.

Reply:
We do need the version chaining in the block though.